lib/Transform/XTenQDQConcat.cpp

```diff
@@ -0,0 +1,67 @@
+#include "xten/Dialect/XTenNN/IR/XTenNNOps.h"
+#include "xten/Transform/Passes.h"
+
+#include <mlir/Dialect/Tosa/IR/TosaOps.h>
+#include <mlir/IR/Matchers.h>
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/Support/LogicalResult.h>
+
+#include <llvm/ADT/SmallVector.h>
+
+using namespace mlir;
+using namespace mlir::tosa;
+using namespace amd::xten_nn;
+
+namespace {
+struct XTenQDQConcat : public OpRewritePattern<DequantizeOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(DequantizeOp op,
+                                PatternRewriter &rewriter) const override {
+    // Match concat->QDQ->concat and remove QDQ, if concats would be foldable.
+    // Try to match an incoming concat
+    Value val;
+    if (!matchPattern(op.getInput(), m_Op<QuantizeOp>(matchers::m_Any(&val)))) {
+      return rewriter.notifyMatchFailure(
+          op, "Dequantize input not produced by Quantize.");
+    }
+
+    auto producer = dyn_cast_or_null<ConcatOp>(val.getDefiningOp());
+    if (producer == nullptr) {
+      return rewriter.notifyMatchFailure(
+          op, "QDQ input not produced by TOSA concat.");
+    }
+
+    uint64_t axis = producer.getAxis();
+
+    bool modified = false;
+    for (Operation *user : op->getUsers()) {
```

Comment:
Should we update some users if we cannot update all users? It seems that in that case the concats cannot be folded.

Then we can we express the check via an `llvm::all_of` and the update via a single `rewriter.replaceOp(op, producer.getOutput()`

Reply:
Since we only update the user and do not otherwise touch the matched quantize node, I don't see a reason why we shouldn't update just some of them to enable at least some folding

Reply:
But a partial rewrite of the QDQ nodes does not enable Concat folding later on in the case where only a part of the users match. So we destroy semantics with no benefit.

Reply:
This is not what is happening here. See the response to other comment on the test case